File: app/src/test/java/com/karthik/myweather/ui/viewModel/SearchCityViewModelTest.java
package com.karthik.myweather.ui.viewModel;

import com.karthik.myweather.Event;
import com.karthik.myweather.Permission;
import com.karthik.myweather.data.WeatherDatabase;
import com.karthik.myweather.navigation.NavEvent;
import com.karthik.myweather.network.WeatherService;
import com.karthik.myweather.network.model.Location;
import com.karthik.myweather.utils.RxScheduler;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class SearchCityViewModelTest extends BaseTest{

    SearchCityViewModel viewModel;

    @Mock
    WeatherDatabase weatherDatabase;

    @Mock
    WeatherService weatherService;

    @Mock
    RxScheduler rxScheduler;

    @Mock
    LocationEntityDao locationEntityDao;

    @Mock
    CityWeatherDao cityWeatherDao;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.initMocks(this);
        when(rxScheduler.io()).thenReturn(Schedulers.trampoline());
        when(rxScheduler.mainThread()).thenReturn(Schedulers.trampoline());
        when(weatherDatabase.locationEntityDao()).thenReturn(locationEntityDao);
        when(weatherDatabase.cityWeatherDao()).thenReturn(cityWeatherDao);
        viewModel = new SearchCityViewModel(weatherService, weatherDatabase, rxScheduler);
    }

    @After
    public void tearDown() throws Exception {
        viewModel = null;
        weatherDatabase = null;
        weatherService = null;
        cityWeatherDao = null;
        locationEntityDao = null;
    }

    @Test
    public void submitQuery() {
        List<Location> locations = new ArrayList<>();
        locations.add(new Location(){{
            title = "title";
            locationId = 99;
        }});
        when(weatherService.getQueryResults(any()))
                .thenReturn(Single.just(locations));
        when(weatherDatabase.locationEntityDao().addAll(anyList()))
                .thenReturn(Completable.complete());
        when(weatherDatabase.locationEntityDao().deleteAll())
                .thenReturn(Completable.complete());
        viewModel.submitQuery();
        verify(locationEntityDao, times(1)).deleteAll();
        verify(locationEntityDao, times(1)).addAll(anyList());
        assertEquals(viewModel.navEvent.getValue(), NavEvent.LocationToCity);
    }

    @Test
    public void submitUseMyLocation() {
        viewModel.submitUseMyLocation();
        assertEquals(viewModel.permissionRequest.getValue(), Event.PermissionRequest);
    }

    @Test
    public void onPermission_Granted() {
        viewModel.onPermission(Permission.GRANTED);
        assertEquals(viewModel.locationRequest.getValue(), Event.LocationRequest);
    }

    @Test
    public void onPermission_Denied() {
        viewModel.onPermission(Permission.DENIED);
        assertNull(viewModel.locationRequest.getValue());
    }

    @Test
    public void getLocationResults() {
        List<Location> locations = new ArrayList<>();
        locations.add(new Location(){{
            title = "title";
            locationId = 99;
        }});
        when(weatherService.getLocationResults(any()))
                .thenReturn(Single.just(locations));
        when(weatherDatabase.locationEntityDao().addAll(anyList()))
                .thenReturn(Completable.complete());
        when(weatherDatabase.locationEntityDao().deleteAll())
                .thenReturn(Completable.complete());
        viewModel.getLocationResults("");
        verify(locationEntityDao, times(1)).deleteAll();
        verify(locationEntityDao, times(1)).addAll(anyList());
        assertEquals(viewModel.navEvent.getValue(), NavEvent.LocationToCity);
    }

    @Test
    public void onLocation() {
        android.location.Location location = mock(android.location.Location.class);
        when(location.getLatitude()).thenReturn(10.0);
        when(location.getLongitude()).thenReturn(11.0);
        SearchCityViewModel viewModelSpy = spy(viewModel);
        doNothing().when(viewModelSpy).getLocationResults(anyString());
        viewModelSpy.onLocation(location);
        verify(viewModelSpy).getLocationResults(anyString());
    }
}

File: app/src/test/java/com/karthik/myweather/ui/viewModel/SearchCityViewModelTest.kt
package com.karthik.myweather.ui.viewModel

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.karthik.myweather.Event
import com.karthik.myweather.Permission
import com.karthik.myweather.data.WeatherDatabase
import com.karthik.myweather.data.dao.CityWeatherDao
import com.karthik.myweather.data.dao.LocationEntityDao
import com.karthik.myweather.navigation.NavEvent
import com.karthik.myweather.network.WeatherService
import com.karthik.myweather.network.model.Location
import com.karthik.myweather.utils.RxScheduler
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import java.util.*

@RunWith(AndroidJUnit4::class)
class SearchCityViewModelTest {
    lateinit var viewModel: SearchCityViewModel

    @Mock
    lateinit var weatherDatabase: WeatherDatabase

    @Mock
    lateinit var weatherService: WeatherService

    @Mock
    lateinit var rxScheduler: RxScheduler

    @Mock
    lateinit var locationEntityDao: LocationEntityDao

    @Mock
    lateinit var cityWeatherDao: CityWeatherDao

    lateinit var closable: AutoCloseable

    @Before
    @Throws(Exception::class)
    fun setUp() {
        closable = MockitoAnnotations.openMocks(this)
        Mockito.`when`(rxScheduler.io()).thenReturn(Schedulers.trampoline())
        Mockito.`when`(rxScheduler.mainThread()).thenReturn(Schedulers.trampoline())
        Mockito.`when`(weatherDatabase.locationEntityDao()).thenReturn(locationEntityDao)
        Mockito.`when`(weatherDatabase.cityWeatherDao()).thenReturn(cityWeatherDao)
        viewModel = SearchCityViewModel(weatherService, weatherDatabase, rxScheduler)
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        closable.close()
    }

    @Test
    fun submitQuery() {
        val locations: MutableList<Location?> = ArrayList()
        locations.add(Location("title", 99))
        Mockito.`when`(weatherService!!.getQueryResults(ArgumentMatchers.any<String>()))
                .thenReturn(Single.just(locations))
        Mockito.`when`(weatherDatabase!!.locationEntityDao()!!.addAll(ArgumentMatchers.anyList()))
                .thenReturn(Completable.complete())
        Mockito.`when`(weatherDatabase!!.locationEntityDao()!!.deleteAll())
                .thenReturn(Completable.complete())
        viewModel!!.submitQuery()
        Mockito.verify(locationEntityDao, Mockito.times(1))!!.deleteAll()
        Mockito.verify(locationEntityDao, Mockito.times(1))!!.addAll(ArgumentMatchers.anyList())
        Assert.assertEquals(viewModel!!.navEvent.value, NavEvent.LocationToCity)
    }

    @Test
    fun submitUseMyLocation() {
        viewModel!!.submitUseMyLocation()
        Assert.assertEquals(viewModel!!.permissionRequest.value, Event.PermissionRequest)
    }

    @Test
    fun onPermission_Granted() {
        viewModel!!.onPermission(Permission.GRANTED)
        Assert.assertEquals(viewModel!!.locationRequest.value, Event.LocationRequest)
    }

    @Test
    fun onPermission_Denied() {
        viewModel!!.onPermission(Permission.DENIED)
        Assert.assertNull(viewModel!!.locationRequest.value)
    }

    @Test
    fun testGetLocations() {
        val locations: MutableList<Location?> = ArrayList()
        locations.add(Location("title", 99))
        Mockito.`when`(weatherService!!.getLocationResults(ArgumentMatchers.any<String>()))
                .thenReturn(Single.just(locations))
        Mockito.`when`(weatherDatabase!!.locationEntityDao()!!.addAll(ArgumentMatchers.anyList()))
                .thenReturn(Completable.complete())
        Mockito.`when`(weatherDatabase!!.locationEntityDao()!!.deleteAll())
                .thenReturn(Completable.complete())
        viewModel!!.getLocationResults("")
        Mockito.verify(locationEntityDao, Mockito.times(1))!!.deleteAll()
        Mockito.verify(locationEntityDao, Mockito.times(1))!!.addAll(ArgumentMatchers.anyList())
        Assert.assertEquals(viewModel!!.navEvent.value, NavEvent.LocationToCity)
    }

    @Test
    fun onLocation() {
        val location = Mockito.mock(android.location.Location::class.java)
        Mockito.`when`(location.latitude).thenReturn(10.0)
        Mockito.`when`(location.longitude).thenReturn(11.0)
        val viewModelSpy = Mockito.spy(viewModel)
        Mockito.doNothing().`when`(viewModelSpy)!!.getLocationResults(ArgumentMatchers.anyString())
        viewModelSpy!!.onLocation(location)
        Mockito.verify(viewModelSpy)!!.getLocationResults(ArgumentMatchers.anyString())
    }
}